### PR TITLE
[FLINK-26217] Introduce manifest.merge-min-count in commit

### PR DESCRIPTION
**Pull request by shenzhu**
Changes in this PR:
- Introduce `manifest.merge-min-count` option in `FileStoreOptions.java`
- Update `merge` function in `ManifestFileMata.java` to take that option

### CODE REVIEW

**Review by JingsongLi — COMMENTED** [review]
Thanks! I will take a look in the next few days

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/manifest/ManifestFileMeta.java [hunk 1]
```diff
@@ -135,19 +135,22 @@ public String toString() {
...
+        int metaCount = 0;
 
         try {
             // merge existing manifests first
             for (ManifestFileMeta manifest : metas) {
                 totalSize += manifest.fileSize;
+                metaCount += 1;
                 candidate.add(manifest);
-                if (totalSize >= suggestedMetaSize) {
+                if (totalSize >= suggestedMetaSize && metaCount >= suggestedMinMetaCount) {
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/manifest/ManifestFileMeta.java:153]: We should choose not to merge when the totalSize is not satisfied and the min count is not satisfied.
So we need check before below `mergeIntoOneFile`.

**Review by JingsongLi — COMMENTED** [review]
BTW, can you add test for this?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/manifest/ManifestFileMeta.java (hunk 1 above)
Comment (shenzhu) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/manifest/ManifestFileMeta.java:153]: > 

Hey @JingsongLi , thanks for your review! Please correct me if I didn't understand it correctly, do you mean we should have this check in `mergeIntoOneFile`?

Comment (shenzhu) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/manifest/ManifestFileMeta.java:153]: > We should choose not to merge when the totalSize is not satisfied and the min count is not satisfied. So we need check before below `mergeIntoOneFile`.

Hey @JingsongLi thanks for your review! Please correct me if I didn't understand it correctly, do you mean we should have this check in `mergeIntoOneFile`?

**Comment by tsreaper** [comment]
Hi @shenzhu ! Firstly thanks for your interest in contributing and also thanks @JingsongLi for the review.

I understand that we're introducing this configuration to optimize merging as well as to decrease the write overhead. However I'd like to suggest a better method for this.

Consider the merging process across snapshots. It's obvious that once a merged file exceeds the size limit it won't be merged again and will stay there forever. That is, the actual merging only happens at the tail of our manifest list. So why not just check the list from tail to head and once we hit a merged file we just exit? In this way we don't need this new configuration and we don't even need to iterate over the whole manifest list.

What do you think?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/manifest/ManifestFileMeta.java (hunk 1 above)
Comment (tsreaper) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/manifest/ManifestFileMeta.java:153]: You're missing the case when we merge the last bit of manifests with entries. See line 172.

**Comment by shenzhu** [comment]
> 

Hey @tsreaper , thanks for your feedback! I think that's a good point

I'm still new to this code base, please correct me if I didn't understand it correctly. I have two questions about that idea

**Q1.** Is there an easy way to know if a manifest file has been merged before?
I was checking the fields in [ManifestFileMeta](https://github.com/apache/flink-table-store/blob/master/flink-table-store-core/src/main/java/org/apache/flink/table/store/file/manifest/ManifestFileMeta.java#L38) but didn't find any indicator

**Q2.** Does order of operations matter?
If we plan to scan from tail to head, then we might see some files deleted first and then added. For example with the following case:
```
merged1, merged2, manifest1(+A), manifest2(+B, +C), manifest3(-B, +D), entry1(-A)
```
The merged results could be something like
```
merged1, merged2, merged3(+A), merged4(+C, +D, -A)
```

**Comment by tsreaper** [comment]
Hi @shenzhu .

After a second thought, I believe we should still stick to your current implementation with a small change. There are several reasons:
* After merging several files into one file, the size of the resulting file should shrink evidently. The file size may drop below the merging threshold again and is eligible for another merge. I stated that after a file is merged we'll never touch it again, but we can see that it is not true.
* If we do not introduce this merging count constraint, we might merge an old large file and a small freshly produced file several times until the old large file exceeds the size threshold. This will cause the write overhead to grow wildly (let's say the old large file is 7MB and it is merged 10 times, then the write overhead will be 7 * 10MB) and this is what we'd like to avoid.

However your current implementation still has some problems.
* Let's begin with a 7MB file. After 29 snapshots (each snapshot produce one 1MB file) you'll merge these 30 files into a 7 + 29 = 36MB file. After another 29 snapshots you'll again merge these 30 files into a 36 + 29 = 65MB file. You can see that the first file will become larger and larger and the write overhead will grow unboundedly. Our current implementation (with size constraint only) does not have this problem.
* You're missing the case when we merge the last bits of files.

To avoid these problems, you should perform merging if either size or count constraint is met. Just like @JingsongLi suggested.

About your question. For Q1 we currently do not have such a mechanism. For Q2 you'll also face this problem in the current implementation. It is OK because manifests are meaningful only if they're viewed as a whole.

**Comment by shenzhu** [comment]
> Hi @shenzhu .
> 
> After a second thought, I believe we should still stick to your current implementation with a small change. There are several reasons:
> 
> * After merging several files into one file, the size of the resulting file should shrink evidently. The file size may drop below the merging threshold again and is eligible for another merge. I stated that after a file is merged we'll never touch it again, but we can see that it is not true.
> * If we do not introduce this merging count constraint, we might merge an old large file and a small freshly produced file several times until the old large file exceeds the size threshold. This will cause the write overhead to grow wildly (let's say the old large file is 7MB and it is merged 10 times, then the write overhead will be 7 * 10MB) and this is what we'd like to avoid.
> 
> However your current implementation still has some problems.
> 
> * Let's begin with a 7MB file. After 29 snapshots (each snapshot produce one 1MB file) you'll merge these 30 files into a 7 + 29 = 36MB file. After another 29 snapshots you'll again merge these 30 files into a 36 + 29 = 65MB file. You can see that the first file will become larger and larger and the write overhead will grow unboundedly. Our current implementation (with size constraint only) does not have this problem.
> * You're missing the case when we merge the last bits of files. That said, these files should not be merged because they do not meet any of the constraints. You should remove this merging and only write the new entries into a manifest file.
> 
> To avoid these problems, you should perform merging if either size or count constraint is met. Just like @JingsongLi suggested.
> 
> About your questions. For Q1 we currently do not have such a mechanism. For Q2 you'll also face this problem in the current implementation. It is OK because manifests are meaningful only if they're viewed as a whole.

Hey @tsreaper , got it, thanks for your explanation!

**Comment by shenzhu** [comment]
Hey @tsreaper , I updated this PR and added some tests, would you mind taking a look again when you have a moment? Thanks!

**Review by tsreaper — APPROVED** [review]
Looks good to me now. @JingsongLi any other thoughts?

**Review by tsreaper — COMMENTED** [review]
Now that entries are written into an independent file, we can remove them from the arguments. New files containing the new entries should be written by the caller of the method.

**Review by JingsongLi — APPROVED** [review]
Thanks @shenzhu and @tsreaper , Looks good to me!

**Pull request merged**